contracts/src/v2/Calls.sol

```diff
@@ -0,0 +1,164 @@
+import {Upgrade} from "../Upgrade.sol";
+import {Functions} from "../Functions.sol";
+import {Constants} from "../Constants.sol";
+
+import {
+    Payload, OperatingMode, Asset, makeNativeAsset, makeForeignAsset, Network
+} from "./Types.sol";
+
+import {UD60x18, ud60x18, convert} from "prb/math/src/UD60x18.sol";
+
+/// @title Library for implementing Ethereum->Polkadot ERC20 transfers.
+library CallsV2 {
+    using Address for address;
+    using SafeTokenTransfer for IERC20;
+    using SafeNativeTransfer for address payable;
+    using TokenInfoFunctions for TokenInfo;
+
+    uint8 public constant MAX_ASSETS = 8;
+
+    // Refer to `IGateway.sendMessage` for documentation
+    function sendMessage(
+        bytes calldata xcm,
+        bytes[] calldata assets,
+        bytes calldata claimer,
+        uint128 executionFee,
+        uint128 relayerFee
+    ) external {
+        _sendMessage(msg.sender, xcm, assets, claimer, executionFee, relayerFee);
+    }
+
+    // Refer to `IGateway.registerToken` for documentation
+    function registerToken(
+        address token,
+        Network network,
+        uint128 executionFee,
+        uint128 relayerFee
+    ) internal {
+        // Build XCM for token registration on AHP and possibly AHK
+        bytes memory xcm;
+
+        require(msg.value <= type(uint128).max, IGatewayV2.ExceededMaximumValue());
+        require(msg.value >= executionFee + relayerFee, IGatewayV2.InsufficientValue());
+        uint128 etherValue = uint128(msg.value) - executionFee - relayerFee;
+
+        if (network == Network.Polkadot) {
+            // Build XCM that executes on AHP only
+            xcm = bytes.concat(
+                hex"deadbeef", abi.encodePacked(token), hex"deadbeef", abi.encodePacked(etherValue)
+            );
+        } else if (network == Network.Kusama) {
+            // Build XCM that is executed on AHP and forwards another message to AHK
+            xcm = bytes.concat(
+                hex"deadbeef", abi.encodePacked(token), hex"deadbeef", abi.encodePacked(etherValue)
+            );
+        } else {
+            revert IGatewayV2.ShouldNotReachHere();
+        }
+
+        Functions.registerNativeToken(token);
+        _sendMessage(address(this), xcm, new bytes[](0), "", executionFee, relayerFee);
+    }
+
+    /*
+    * Internal functions
+    */
+
+    function _sendMessage(
+        address origin,
+        bytes memory xcm,
+        bytes[] memory assets,
+        bytes memory claimer,
+        uint128 executionFee,
+        uint128 relayerFee
+    ) internal {
+        // Ensure outbound messaging is allowed
+        _ensureOutboundMessagingEnabled();
+
+        require(msg.value <= type(uint128).max, IGatewayV2.ExceededMaximumValue());
+        require(msg.value >= executionFee + relayerFee, IGatewayV2.InsufficientValue());
+        address assetHubAgent = Functions.ensureAgent(Constants.ASSET_HUB_AGENT_ID);
+        payable(assetHubAgent).safeNativeTransfer(msg.value);
```

Comment:
This one line does a lot, it essentially takes the fee, reward and any locked eth and passes it to the asset hub agent. Maybe we should collect fees and reward in a separate account like the gateway proxy contract address and just leave locked funds in the asset hub agent.

Reply:
Well for E->P flow in V2, fees and rewards are both ultimately redeemable as WETH (or native ETH) on AH. So they are actually locked funds.